From 08a8a93ba6ef6bf606f7fe7acaf69c5b7c2bcce6 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:49:12 -0400
Subject: [fs] nfsd: -EINVAL on invalid anonuid/gid instead of silent failure

Message-id: <1403560314-32525-11-git-send-email-bfields@redhat.com>
Patchwork-id: 84464
O-Subject: [Patch RHEL7 010/172] nfsd: -EINVAL on invalid anonuid/gid instead of silent failure
Bugzilla: 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

If we're going to refuse to accept these it would be polite of us to at
least say so....

This introduces a slight complication since we need to grandfather in
exportfs's ill-advised use of -1 uid and gid on its test_export.

If it turns out there are other users passing down -1 we may need to
do something else.

Best might be to drop the checks entirely, but I'm not sure if other
parts of the kernel might assume that a task can't run as uid or gid -1.

Cc: "Eric W. Biederman" <ebiederm@xmission.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 6f6cc3205c5f10129b8a10cdf8abf85d9db48a60
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/export.c b/fs/nfsd/export.c
index af51cf9..8513c59 100644
--- a/fs/nfsd/export.c
+++ b/fs/nfsd/export.c
@@ -580,16 +580,25 @@ static int svc_export_parse(struct cache_detail *cd, char *mesg, int mlen)
   if (err)
    goto out4;
   /*
+   * No point caching this if it would immediately expire.
+   * Also, this protects exportfs's dummy export from the
+   * anon_uid/anon_gid checks:
+   */
+  if (exp.h.expiry_time < seconds_since_boot())
+   goto out4;
+  /*
    * For some reason exportfs has been passing down an
    * invalid (-1) uid & gid on the "dummy" export which it
    * uses to test export support.  To make sure exportfs
    * sees errors from check_export we therefore need to
    * delay these checks till after check_export:
    */
+  err = -EINVAL;
   if (!uid_valid(exp.ex_anon_uid))
    goto out4;
   if (!gid_valid(exp.ex_anon_gid))
    goto out4;
+  err = 0;
  }
 
  expp = svc_export_lookup(&exp);
-- 
1.7.1